### PR TITLE
Add gradient registration and tests for Min/Max

### PR DESCRIPTION
**Pull request by mindest**
**Description**: Add gradient registration and corresponding tests for ops `Min` and `Max`

**Motivation and Context**
- Support training of a model that uses `Max`
- `Min` shares the same logic as `Max`

### CODE REVIEW

**Comment by mindest** [comment]
In max gradient calculation, the gradient should be given to the max input.

When there are multiple maxima, there are also multiple ways to deal with it, some of which are:

1. assign gradient to all the max inputs, 1 for each. [A previous PR on Min/Max gradient](https://github.com/microsoft/onnxruntime/pull/6286) follows this.
2. assign gradient to only one of the max inputs, i.e., only one of them gets 1, others get 0.
3. evenly distribute gradient among the max inputs, i.e., for _n_ maxima, each max input gets a gradient of 1/_n_.

image

In a recent case of TwinBERT training (in fp16), it was found that the first implementation may cause some weights to diverge (the divergence seemed to be accumulating step by step). In contrast, the latter two both worked fine. Therefore, keeping a total gradient of 1 should be more reasonable, at least for our TwinBERT training case. Some more related discussion can be found [here](https://github.com/tensorflow/tensorflow/issues/4886).

Specially for this PR, the second implementation is adopted -- let the last max input get the gradient 1 and all others gradient 0.

**Comment by SherlockNoMad** [comment]
Hi @mindest, out of 3 approaches, do you know which approach pytorch is using? We should match pytorch's numerical result in this case.

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 1]
```diff
@@ -2423,6 +2423,86 @@ TEST(GradientCheckerTest, ClipGrad) {
...
+    TensorInfo x1_info({2, 3, 4}, true);
+    TensorInfo x2_info({3, 4}, true);
+    TensorInfo x3_info({4}, true);
+    TensorInfo y_info({2, 3, 4}, true);
+    gradient_checker.ComputeGradientError(op_def, {x1_info, x2_info, x3_info}, {y_info}, &max_error);
+    EXPECT_IS_TINY(max_error);
+  }
+}
+
+TEST(GradientCheckerTest, MaxGrad) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:2466]: MinGrad and MaxGrad can share the same helper function here.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 2]
```diff
@@ -1502,5 +1502,91 @@ IMPLEMENT_GRADIENT_BUILDER(GetAbsGradient) {
   };
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetMinMaxGradient) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1505]: Hmm... this gradient involves a lot of small elementwise computation... 
I am wondering if the Max Op has an addition output, i.e. the index of selected input, can help reducing the amount of computation. 
Just like the MaxPool's second output.

Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1505]: Those small element wise ops are mainly to handle the case when number of inputs are greater than 2, for number of inputs is 2 (which is the common case, expecially when using PyTorch, whose Min/Max supports only 2 arguments), the code can generate the optimized sub-graph, it's Equal->Cast->Mul for one input, and Not->Cast->Mul for the other one.
If we add the indices output, do we need to start the change from ONNX? And I think the optimized way to use the indices output in backward is to add and write the kernel for Min/MaxGrad Ops, just like MaxPoolGrad. If not, i.e., still use existing Ops, for each input, it will still need Equal->Cast->Mul. The question is, how much perf gain we can get by doing this?

**Comment by mindest** [comment]
> Hi @mindest, out of 3 approaches, do you know which approach pytorch is using? We should match pytorch's numerical result in this case.

In PyTorch (torch1.8.1), when there are two inputs for `torch.max`, the gradient will be granted to the latter one if both are max:
```py
import torch
a = torch.tensor((1., 2., 3.), requires_grad=True)
b = torch.tensor((3., 2., 1.), requires_grad=True)
y = torch.max(a, b)
print(y) # tensor([3., 2., 3.], grad_fn=<MaximumBackward>)
y.backward(gradient=torch.Tensor([1, 1, 1]))
print(a.grad) # tensor([0., 0., 1.])
print(b.grad) # tensor([1., 1., 0.])
```
PyTorch max does not support over 2 inputs, but it can accept one input, and compute the max along the axis, where there can be multiple values to compare. I think this works like `ReduceMax` in onnx. As I tried, there are two ways:
```py
c = torch.tensor((1., 2., 2., 1.), requires_grad=True)
d = torch.tensor((1., 2., 2., 1.), requires_grad=True)

p = torch.max(c, 0)
print(p.values) # tensor(2., grad_fn=<MaxBackward0>)
p.values.backward()
print(c.grad) # tensor([0., 1., 0., 0.])

q = torch.max(d)
print(q) # tensor(2., grad_fn=<MaxBackward1>)
q.backward()
print(d.grad) # tensor([0.0000, 0.5000, 0.5000, 0.0000])
```
i.e., MaxBackward0 gives gradient to the first max input; MaxBackward1 distributes gradient among the max inputs. These results are in accordance with generated autograd function codes of PyTorch.

In onnx, the definition of `Max` is different from that of PyTorch. It can accept multiple inputs: 
- when there is only one input, it just returns the input;
- when there are two or more inputs, it does elementwise comparison.

Here I chose to match its behavior with PyTorch for two inputs (i.e., give gradient to the latter one for two-max case, and to the last one for multiple-max case). But I'm not sure whether it should be done like this for over two inputs, for there seems no PyTorch equivalent to refer to.

**Comment by SherlockNoMad** [comment]
Thanks for the detailed explanation here! As you have showed, there are two gradient functions for Max: MaxBackward0 and MaxBackward1, and they behave differently.  

In order to match the numerical result of pytorch, we should distinguish these two Max ops, potentially we can add an attribute "gradient_mode" to the onnx's Max op. 

 Please open an discussion with Spandan and Ashwini. Exporter will need to distinguish these two Max, onnx op should support this.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1505]: I don't want to block this PR's merge, but it seems we need more time to figure out the gradient behavior for more than 2 inputs. 

How about we limit the gradient builder to only support 2 input case? It can ORT_THROW for the other cases?

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 3]
```diff
@@ -2423,6 +2423,54 @@ TEST(GradientCheckerTest, ClipGrad) {
   }
 }
 
+void GradientCheckerMinMaxGradHelper(std::string op) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:2426]: pass by const ref

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 4]
```diff
@@ -2423,6 +2423,54 @@ TEST(GradientCheckerTest, ClipGrad) {
...
+  float max_error;
+  GradientChecker<float, float, float> gradient_checker;
+  OpDef op_def{op, kOnnxDomain, 11};
+
+  {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:2431]: need one case for matching values

Comment (mindest) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:2431]: Does "matching values" here mean two inputs have the same value? If so, the third test case has some specified equal values. - Is it enough for this purpose?

Comment (SherlockNoMad) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:2431]: yes

Comment (mindest) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:2431]: Test case 3 was meant to have some equal values for the two inputs, but I made a mistake here -- did not pass the x_datas.

When correcting the mistake, I realized that Max/Min is not differentiable at the equal value, and the test for equal case will just fail because the calculated numeric Jacobian and theoretical Jacobian are different. I discussed with @iK1D and think that maybe we can replace the "wrong" numeric Jacobian with the expected Jacobian (we know it for a specific input), and compare it with the theoretical one. Is this ok? 

Or should we just avoid the equal case? I see that Relu is similar, and ReluGrad's test does not have case for 0, where it's not differentiable.

Comment (SherlockNoMad) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:2431]: ok... since we know why the UT is failing, let's just skip the equal case for now. 
please add a comment to document this UT case though...

**Review by SherlockNoMad — APPROVED** [review]
